src/crimson/osd/pg.cc

```diff
@@ -743,37 +745,58 @@ PG::do_osd_ops(
-    }
-    auto reply = make_message<MOSDOpReply>(m.get(),
-                                           result,
-                                           get_osdmap_epoch(),
-                                           0,
-                                           false);
-    reply->add_flags(CEPH_OSD_FLAG_ACK | CEPH_OSD_FLAG_ONDISK);
-    logger().debug(
-      "do_osd_ops: {} - object {} sending reply",
-      *m,
-      obc->obs.oi.soid);
-    return PG::do_osd_ops_ertr::make_ready_future<Ref<MOSDOpReply>>(
-      std::move(reply));
+  }).safe_then_unpack_interruptible_tuple(
+    [this, m, obc, rvec = op_info.allows_returnvec(), ox = ox.get()]
+    (auto submitted_fut, auto all_completed_fut)
+    -> PG::do_osd_ops_iertr::future<rep_op_fut_t> {
+    return seastar::make_ready_future<rep_op_fut_t>(
+          std::move(submitted_fut),
+          all_completed_fut.safe_then_interruptible_tuple(
+          [this, m, obc, rvec]()
+          -> PG::do_osd_ops_iertr::future<Ref<MOSDOpReply>> {
+            // TODO: should stop at the first op which returns a negative retval,
+            //       cmpext uses it for returning the index of first unmatched byte
+            int result = m->ops.empty() ? 0 : m->ops.back().rval.code;
+            if (result > 0 && !rvec) {
+              result = 0;
+            }
+            auto reply = make_message<MOSDOpReply>(m.get(),
+                                                   result,
+                                                   get_osdmap_epoch(),
+                                                   0,
+                                                   false);
+            reply->add_flags(CEPH_OSD_FLAG_ACK | CEPH_OSD_FLAG_ONDISK);
+            logger().debug(
+              "do_osd_ops: {} - object {} sending reply",
+              *m,
+              obc->obs.oi.soid);
+            return seastar::make_ready_future<Ref<MOSDOpReply>>(
+                std::move(reply));
+          }, crimson::ct_error::object_corrupted::handle([m, obc, this] {
+            return repair_object(m, obc->obs.oi.soid, obc->obs.oi.version)
+            .then_interruptible([] {
+              return PG::do_osd_ops_ertr::future<Ref<MOSDOpReply>>(
+                      crimson::ct_error::eagain::make());
+            });
+          }), osd_op_errorator::all_same_way([ox,
+                                              m,
+                                              obc,
+                                              this] (const std::error_code& e) {
+            return handle_failed_op(e, std::move(obc), *ox, *m);
+          }))
+        );
   }, crimson::ct_error::object_corrupted::handle([m, obc, this] {
     return repair_object(m, obc->obs.oi.soid, obc->obs.oi.version)
     .then_interruptible([] {
-      return PG::do_osd_ops_ertr::future<Ref<MOSDOpReply>>(
-      crimson::ct_error::eagain::make());
+      return PG::do_osd_ops_ertr::future<rep_op_fut_t>(
+              crimson::ct_error::eagain::make());
```

Comment:
does it miss a seastar::now() future here? since return rep_op_fut_t it is a tuple。

Reply:
please note, we are not creating a future with its value here. we are returning an future with error. in this case, we don't pass a tuple or the argument for constructing a tuple to the constructor of `errorator::future<>`. otherwise, we'd use `make_ready_future<>()` instead.